test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/ArrayOutput.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Sockets.Tests.Internal.Formatters
+{
+    internal class ArrayOutput : IOutput
+    {
+        private IList<ArraySegment<byte>> _buffers = new List<ArraySegment<byte>>();
+
+        private int _chunkSize;
+        private byte[] _activeBuffer;
+        private int _offset;
+
+        public Span<byte> Buffer => _activeBuffer.Slice(_offset);
+
+        public ArrayOutput(int chunkSize)
+        {
+            _chunkSize = chunkSize;
+            AdvanceChunk();
+        }
+
+        public void Advance(int bytes)
+        {
+            // Determine the new location
+            _offset += bytes;
+            Debug.Assert(_offset <= _activeBuffer.Length, "How did we write more data than we had space?");
+        }
+
+        public void Enlarge(int desiredBufferLength = 0)
+        {
+            if (desiredBufferLength == 0 || _activeBuffer.Length - _offset < desiredBufferLength)
```

Comment:
Since the caller can see the current buffer size, this indicates that the current buffer isn't large enough, so we just make a new one. This (test) implementation is designed to use small fixed-size buffers to ensure we handle writing to buffer chains properly, so we just grab a new one when `desiredBufferLength` is `0`.